webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartling.java

```diff
@@ -123,9 +132,19 @@ public ThirdPartyTMSSmartling(SmartlingClient smartlingClient,
     @Override
     public ThirdPartyTMSImage uploadImage(String projectId, String name, byte[] content) {
         logger.debug("Upload image to Smartling, project id: {}, name: {}", projectId, name);
-        ContextUpload contextUpload = smartlingClient.uploadContext(projectId, name, content);
+        Optional<ContextUpload> contextUpload = Mono.fromCallable(() -> smartlingClient.uploadContext(projectId, name, content))
+                .retryWhen(getRetryConfiguration()
+                        .doBeforeRetry(e -> logger.info("Retrying after image upload to Smartling failed, project id: {}, name: {}, error: {}", projectId, name, e.failure())))
+                .doOnError(e -> {
+                    String msg = String.format("Error uploading image to Smartling, project id: %s, name: %s", projectId, name);
+                    logger.error(msg, e);
+                    throw new SmartlingClientException(msg, e);
+                })
+                .blockOptional();
+        contextUpload.orElseThrow(() -> new SmartlingClientException("Error with image upload to Smartling, context upload information is not present."));
+
         ThirdPartyTMSImage thirdPartyTMSImage = new ThirdPartyTMSImage();
-        thirdPartyTMSImage.setId(contextUpload.getContextUid());
+        thirdPartyTMSImage.setId(contextUpload.get().getContextUid());
```

Comment:
you can move the `.orElseThrow(` after `blockOptional` and keep `ContextUpload` instead of `Optional<ContextUpload>`

and in fact you probably even do a `map` that instantiate the  `ThirdPartyTMSImage`